### PR TITLE
Initial changes to try using buildtools v2 for WCF

### PR DESCRIPTION
**Pull request by StephenBonikowsky**
@dagood this PR contains all my initial changes, I compared file by file everything in CoreFx and pulled everything across except things related to native code as WCF doesn't have any.

My first step was to just try to msbuild System.Private.ServiceModel.csproj, I've been looking at [System.Private.Xml](https://github.com/dotnet/corefx/tree/master/src/System.Private.Xml) in CoreFx as a guide.
I confirmed that I could msbuild that project with args, 
> /p:ConfigurationGroup=Release /p:TargetOS=Windows_NT

Msbuilding S.P.SM is not working as it isn't pulling in any of the assembly references I added in the .csproj. So trying to figure out why not is what I am currently trying to figure out. My next step was to add a simpler project with just enough source code to cause a couple of compilation failures instead of the 46K I'm getting with S.P.SM and use that to investigate further.

I'm having to work on some other things today, if you have any ideas let me know.

### CODE REVIEW

**Comment by StephenBonikowsky** [comment]
Adding @MattGal 
I created a simple project ad what I am finding is that my packages folder is empty except for...

\WCF\packages\microsoft.dotnet.buildtools

So regardless of what references I have in my .csproj they aren't getting pulled down and copied into the bin folder.

**Comment by MattGal** [comment]
@StephenBonikowsky will take a look in a bit and let you know if I see anything useful to suggest.

**Comment by StephenBonikowsky** [comment]
In the MSBuild log I found...

`Task "GetReferenceAssemblyPaths" skipped, due to false condition; ('$(TargetFrameworkMoniker)' != '' and ('$(_TargetFrameworkDirectories)' == '' or '$(_FullFrameworkReferenceAssemblyPaths)' == '')) was evaluated as ('.NETPortable,Version=v4.5' != '' and ('C:\OSS\WCF\Tools\' == '' or 'C:\OSS\WCF\Tools\' == '')).`

It doesn't seem correct that $(TargetFrameworkMoniker) was evaluated as '.NETPortable,Version=v4.5'

**Comment by MattGal** [comment]
I think the first thing to do here is to work through the failures in sync.cmd, as that's likely related to your un-populated packages folder and necessary for other stuff to work.  Using this branch, the first time I run sync I get:

```
C:\GIT\wcf\Tools\OptionalTooling.targets(50,5): error MSB3030: Could not copy the file "C:\GIT\wcf\\external\test-runtime\optional.json" because it was not found. [C:\GIT\wcf\build.proj]
```

... and disabling that by not running it if the file is missing I fail in EncryptedConfigNuGetRestore.   Getting buildtools and WCF to play nicely on this step is a good first step.
(If I'm missing something about how to make this actually work, please let me know)

File: dir.props [hunk 1]
```diff
@@ -559,29 +192,69 @@
...
+    <!-- Constructed shared fx path for testing -->
+    <UseDotNetNativeToolchain Condition="'$(BuildingUAPAOTVertical)' == 'true'">true</UseDotNetNativeToolchain>
 
     <PackagesBasePath Condition="'$(PackagesBasePath)'==''">$(BinDir)$(OSPlatformConfig)</PackagesBasePath>
     <PackageOutputPath Condition="'$(PackageOutputPath)'==''">$(PackageOutputRoot)$(ConfigurationGroup)/</PackageOutputPath>
     <SymbolPackageOutputPath Condition="'$(SymbolPackageOutputPath)'==''">$(PackageOutputPath)symbols/</SymbolPackageOutputPath>
   </PropertyGroup>
 
   <PropertyGroup>
+    <OptionalToolingJsonPath>$(ProjectDir)\external\test-runtime\optional.json</OptionalToolingJsonPath>
```
Comment (dagood) [inline dir.props]: Removing this `PropertyGroup` entirely will use the default optional tooling json path, which should fix the `EncryptedConfigNuGetRestore` Matt's seeing. But I'm not sure why the optional tooling restore is enabled at all--it shouldn't run unless specifically configured, and I don't think WCF needs any of the optional tooling.

**Comment by StephenBonikowsky** [comment]
I have sync working now and packages (more than needed) are getting pulled down now.

Next issue I'm working on is that when resolving references it either isn't looking in the correct folder or the incorrect folder is getting created.

Building the very simple System.Private.ServiceModel - Copy.csproj with only a reference to System.Runtime...

```
ResolveAssemblyReferences:
  Primary reference "System.Runtime".
C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1820,5): warning MSB3245: Could not resolve this reference. Could not locate the assembly "System.Runtime". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compilation errors. [C:\OSS\WCF\src\System.Private.ServiceModel\src\System.Private.ServiceModel - Copy.csproj]
          For SearchPath "C:\OSS\WCF\bin/ref/netcoreapp/".
          Considered "C:\OSS\WCF\bin/ref/netcoreapp/System.Runtime.winmd", but it didn't exist.
          Considered "C:\OSS\WCF\bin/ref/netcoreapp/System.Runtime.dll", but it didn't exist.
          Considered "C:\OSS\WCF\bin/ref/netcoreapp/System.Runtime.exe", but it didn't exist.
          For SearchPath "{RawFileName}".
          Considered treating "System.Runtime" as a file name, but it didn't exist.
```

But System.Runtime is actually located...
```
C:\OSS\WCF\bin\ref\netfx
C:\OSS\WCF\bin\ref\netstandard
```
So should it be finding a "C:\OSS\WCF\bin/ref/netcoreapp" folder instead of "C:\OSS\WCF\bin\ref\netstandard"?

**Comment by StephenBonikowsky** [comment]
@MattGal and @dagood 
I'm well past my prior comment, slowly making progress. My current issue is that after adding ILLink from the externals directory it does...
```
Creating directory "C:\OSS\WCF\bin/obj/AnyOS.AnyCPU.Release/ConsoleApplication1/netstandard/PreTrim/".
Moving file from "C:\OSS\WCF\bin/obj/AnyOS.AnyCPU.Release/ConsoleApplication1/netstandard/ConsoleApplication1.dll" to "C:\OSS\WCF\bin/obj/AnyOS.AnyCPU.Release/ConsoleApplication1/netstandard/PreTrim/ConsoleApplication1.dll".
```
After which my build fails because...
`C:\OSS\WCF\Tools\sign.targets(110,5): error : C:\OSS\WCF\bin/obj/AnyOS.AnyCPU.Release/ConsoleApplication1/netstandard/ConsoleApplication1.dll: I/O error reading or writing PE file: Could not find file 'C:\OSS\WCF\bin\obj\AnyOS.AnyCPU.Release\ConsoleApplication1\netstandard\ConsoleApplication1.dll'. [C:\OSS\WCF\src\System.Private.ServiceModel\src\ConsoleApplication1.csproj]`

I'm not sure what the purpose of ILLink is, do you guys know?

**Comment by dagood** [comment]
I found this commit that describes what ILLink is doing in CoreFX, but I haven't seen it before: https://github.com/dotnet/corefx/commit/c42fd5602476dc88fd0d677a23ed15e8592a91e9. If that doesn't sound like something WCF needs, leaving it out (for now?) sounds reasonable to me. (Assuming it doesn't break things to leave it out.)

My guess would be that the ILLink targets are partially running--moving the DLL out into the "pre-process" dir but not actually processing it (or putting the output somewhere wrong) so later targets like signing can't find the DLL.

**Comment by StephenBonikowsky** [comment]
I can now sync and msbuild the test project successfully by passing in `/p:ILLinkTrimAssembly=false`
I am now working on building S.P.SM, it has dependencies on some packages that are not part of netstandard such as System.Security.Principal.Windows, trying to figure out how to pull them in successfully.

**Comment by StephenBonikowsky** [comment]
@ericstj I am referencing System.Security.Principal.Windows as I seen done in some CoreFx projects and the project is looking for it...

```
C:\Program Files (x86)\MSBuild\14.0\bin\Microsoft.Common.CurrentVersion.targets(1820,5): warning MSB3245: Could not resolve this reference. Could not locate th
e assembly "System.Security.Principal.Windows". Check to make sure the assembly exists on disk. If this reference is required by your code, you may get compila
tion errors. [C:\OSS\WCF\src\System.Private.ServiceModel\src\System.Private.ServiceModel.csproj]
          For SearchPath "C:\OSS\WCF\bin/ref/netstandard/".
          Considered "C:\OSS\WCF\bin/ref/netstandard/System.Security.Principal.Windows.winmd", but it didn't exist.
          Considered "C:\OSS\WCF\bin/ref/netstandard/System.Security.Principal.Windows.dll", but it didn't exist.
          Considered "C:\OSS\WCF\bin/ref/netstandard/System.Security.Principal.Windows.exe", but it didn't exist.
          For SearchPath "{RawFileName}".
          Considered treating "System.Security.Principal.Windows" as a file name, but it didn't exist.
```
But while it has been pulled down into the packages folder during sync it isn't getting bin placed to 

> C:\OSS\WCF\bin/ref/netstandard

, can you tell me how I get the pre-build process to properly bin place this package?

**Comment by ericstj** [comment]
You'll want to add that to https://github.com/dotnet/wcf/pull/1960/files#diff-77eeb825299844298e2033f534af39efR20 with the appropriate condition.  That package is not part of netstandard and if you want to enable projects in your repo to use it you need to opt-in.  Be aware that package is not part of NETStandard and is not supported on all runtimes, on Unix it will throw PlatformNotSupportedException at runtime.

**Comment by StephenBonikowsky** [comment]
Thank you @ericstj what property would you recommend using for the condition? We had been using $(TargetsWindows) but the updated dir.props that I pulled down from CoreFx no longer has [this section](https://github.com/dotnet/wcf/blob/master/dir.props#L251).

After adding it back in it still wasn't honored and I had to pass it in as an arg in order for the build of S.P.SM to work with it.

**Comment by ericstj** [comment]
TargetsWindows is determined by configuration you're building your library for.  Even so that's not needed to fix this issue, I was just cautioning you about using that library since it'll throw on Unix.

To fix the original issue you asked about you need to add
```xml
<PackageReference Condition="'$(NETStandardVersion)' &gt;= 1.3" Include="System.Security.Principal.Windows">
  <Version>latest-corefx-version</Version>
</PackageReference>
```
to netstandard.depproj.

That said, this whole PR seems a bit busted to me.  I thought WCF was going to move to building more on the public tools rather than relying more on buildtools?

If you do go this route you should really make sure you have a better understanding of the purpose of the stuff you're copying over.  There are plenty of things in external you either don't need or need to change in order to make appropriate for WCF (harvest packages should have WCF packages, you need something to bring in corefx bits for NETCoreapp since you aren't building those, etc).  /cc @weshaggard

**Comment by StephenBonikowsky** [comment]
There seems to be some miscommunication going on. This PR is part of WCF depending on buildtools v2 which is what I believe was discussed in several meetings/email threads.

@zhenlan perhaps you can help clarify the story?

**Comment by zhenlan** [comment]
> That said, this whole PR seems a bit busted to me.  I thought WCF was going to move to building more on the public tools rather than relying more on buildtools?

@ericstj, yes, using public tools was one of the options that was considered briefly. However, since WCF shares so much the same infrastructure with corefx/coreclr, for example, CI, Helix, VSTS, MC, publishing etc. and all of these have been heavily customized with "private" tools, so using public tools is not really an option (unless we reinvent the whole thing ourselves).

> If you do go this route you should really make sure you have a better understanding of the purpose of the stuff you're copying over...

Agreed! We need someone who has more expertise for this work and @dleeapho and @dagood from build tools team have agreed to help (tracked in dotnet/core-eng#793), but they are busy at the moment, so @StephenBonikowsky is trying to assist to make baby step progress. Fortunately, @dleeapho just told me @dagood will make this his priority very soon (thanks!).

@ericstj, any suggestions for @dagood and @StephenBonikowsky to look into as the next step will be helpful and appreciated!

**Comment by ericstj** [comment]
As far as suggestions go, please look through everything in external and make sure it looks right for WCF.  Each of those projects has a purpose and is meant to pull in dependencies specific to the repos position in the stack.

**Pull request merged**